tool/teleport/common/sftp.go

```diff
@@ -75,17 +79,96 @@ func (c compositeCh) Close() error {
 	return trace.NewAggregate(c.r.Close(), c.w.Close())
 }
 
+type byteReader struct {
+	buf []byte
+	io.Reader
+}
+
+func newByteReader(r io.Reader) *byteReader {
+	return &byteReader{
+		Reader: r,
+		buf:    make([]byte, 1),
+	}
+}
+
+func (r *byteReader) ReadByte() (byte, error) {
+	_, err := r.Reader.Read(r.buf)
+	if err != nil {
+		return 0, err
+	}
+
+	return r.buf[0], nil
+}
+
+type allowedOps struct {
+	write bool
+	path  string
+}
+
 // sftpHandler provides handlers for a SFTP server.
 type sftpHandler struct {
-	logger *log.Entry
-	events chan<- *apievents.SFTP
+	logger  *log.Entry
+	allowed *allowedOps
+	events  chan<- *apievents.SFTP
 }
 
-func newSFTPHandler(logger *log.Entry, events chan<- *apievents.SFTP) *sftpHandler {
+func newSFTPHandler(logger *log.Entry, req *srv.FileTransferRequest, events chan<- *apievents.SFTP) (*sftpHandler, error) {
+	var allowed *allowedOps
+	if req != nil {
+		allowed = &allowedOps{
+			write: !req.Download,
+		}
+		// TODO(capnspacehook): reject relative paths and symlinks
```

Comment:
When do we plan to address this TODO?

Reply:
Soon, it's tracked here: https://github.com/gravitational/teleport-private/issues/1401